doc/design/error_reporting.md

```diff
@@ -0,0 +1,456 @@
+
+```cpp
+int x;
+// ...
+IOX_REPORT_IF(x<0, Code::OutOfBounds, RUNTIME_ERROR);
+```
+
+Fatal errors can be conditionally reported in a similar way.
+
+```cpp
+IOX_REPORT_FATAL_IF(x<0, Code::OutOfBounds);
+```
+
+### Require That a Condition Holds
+
+Similarly we can conditionally check whether a condition does hold and report a fatal error in
+the case that it does not hold
+
+```cpp
+int x;
+// ...
+IOX_REQUIRE(x>=0, Code::OutOfBounds);
+```
+
+The condition is required to hold and this requirement is always checked.
+If the condition does not hold, panic is invoked and the execution stops.
+
+This should be used for conditions that may not hold on the correct path, e.g. for error cases.
+It should not be used for assumptions that have to be true in correct code 
+(use `IOX_ASSUME` or `IOX_PRECONDITION` for this).
+
+Note that no condition can generally be enforced in the sense that it must be true and no checking is required.
+
+## Using the API to Check Contracts and Assumptions
+
+The following checks can be disabled and are intended to increase safety during incorrect
+use, specifically detect incorrect use at runtime.
+
+This means they should only be used for checks that are not needed in correct code (i.e. defensive
+programming).
+
+If these checks are disabled, there is no overhead in the code, i.e. no checking or reporting takes place.
+
+### Checking Preconditions
+
+A precondition check
+
+```cpp
+int f(int x)
+{
+IOX_PRECONDITION(x>=0, "precondition violation message");
+
+// ...
+}
+```
+
+is used to verify assumptions **BEFORE** any logic in the function body is executed. Technically copy
+constructors may run before any condition can be checked, and there is also the possibility of
+reordering if the following code does not depend on the condition at all. 
+This is not a problem since any reordering is not allowed to affect the observable result. 
+Specifically it cannot affect the value of the precondition itself as this would change the
+observable behaviour.
+
+In case of violation, the violation and a (potentially empty) message are forwarded to the backend,
+panic is invoked and execution stops.
+
+The verification can be optionally disabled, and hence this also documents assumptions of the
+function itself.
+
+### Checking Assumptions
+
+Checking assumptions is similar to checking preconditions, but can happen anywhere in the code.
+
+```cpp
+int f(int x)
+{
+    // some computation
+    int y = g(x);
+
+    IOX_ASSUME(y>=0, "assumption violation message");
+    // proceed assuming that y>=0 holds
+}
+```
+
+These serve as documentation of assumptions that should hold at this point in the code before the
+next statement and can be used e.g. to check for out of bounds accesses. It can also be used to
+check postconditions.
+
+It should not be used at the start of a function body and instead replaced with a precondition check
+in this case.
+
+### Marking Unreachable Code
```

Comment:
This should not be under a section which states that it can be disabled at compile time, like `## Using the API to Check Contracts and Assumptions` indicates

Reply:
Put into its own section (reduced header indent level).